src/GafferArnold/IECoreArnoldPreview/Renderer.cpp

```diff
@@ -3368,8 +3370,39 @@ class ArnoldGlobals
 		// we have to make do with a global handler.
 		static void aiMsgCallback( int logmask, int severity, const char *msg_string, int tabs )
 		{
-			g_currentMessageHandler->handle( g_ieMsgLevels[ min( severity, 3 ) ], "Arnold", std::string( tabs, ' ' ) + msg_string );
+			const IECore::Msg::Level level = \
+				( logmask == AI_LOG_DEBUG ) ? IECore::Msg::Level::Debug : g_ieMsgLevels[ min( severity, 3 ) ];
+
+			std::stringstream msg;
+
+			const int flags = AiMsgGetConsoleFlags();
+			if( flags & AI_LOG_TIMESTAMP )
+			{
+				const boost::posix_time::time_duration elapsed = boost::posix_time::millisec( AiMsgUtilGetElapsedTime() );
+				msg << std::setfill( '0' );
+				msg << std::setw( 2 ) << elapsed.hours() << ":";
+				msg << std::setw( 2 ) << elapsed.minutes() << ":";
+				msg << std::setw( 2 ) << elapsed.seconds() << " ";
+			}
+			if( flags & AI_LOG_MEMORY )
+			{
+				const size_t mb = AiMsgUtilGetUsedMemory() / 1024 / 1024;
+				if( mb < 1024 )
+				{
+					msg << mb << "MB  ";
+				}
+				else
+				{
+					msg.setf( std::ios::fixed, std:: ios::floatfield );
+					msg << setprecision( 1 ) << ( float( mb ) / 1024.0f ) << "GB ";
+				}
+			}
+
+			msg << std::string( tabs, ' ' ) << msg_string;
```

Comment:
C++ people, let me know if there preferred ways to do all this...

Reply:
I wish I knew! My instinct was to look for something in `std::chrono`, but that doesn't seem to support formatting until C++20, when we finally get `std::format()`. Until then I'd maybe be tempted to use `boost::format()` since it's one step closer to `std::format()` and I find all these stream operators pretty unreadable. Not sure I feel strongly enough about it to change things now though...

Reply:
Its a fun journey isn't it! I wondered if the `stringstream` modifiers might be a little more performant, but its probably us at most...